### PR TITLE
Rewrite various resolve/diagnostics errors as translatable diagnostics

### PR DESCRIPTION
**Pull request by NotStirred**
additional question:

For trivial strings is it ever accepted to use `fluent_generated::foo` in a `label` for example? Or is an empty struct `Diagnostic` preferred?

### CODE REVIEW

File: compiler/rustc_resolve/src/late.rs [hunk 1]
```diff
@@ -2244,12 +2245,11 @@ impl<'a: 'ast, 'b, 'ast, 'tcx> LateResolutionVisitor<'a, 'b, 'ast, 'tcx> {
...
-                let what = if ns == TypeNS { "type parameters" } else { "local variables" };
                 if this.should_report_errs() {
-                    this.r
-                        .tcx
-                        .sess
-                        .span_err(ident.span, format!("imports cannot refer to {}", what));
+                    let what = if ns == TypeNS { "type parameters" } else { "local variables" };
+
+                    let err = ImportsCannotReferTo { span: ident.span, what };
+                    this.r.tcx.sess.create_err(err).emit();
```
Comment (WaffleLapkin) [inline compiler/rustc_resolve/src/late.rs:2252]: Is there a reason not to
```suggestion
                    let what = if ns == TypeNS { "type parameters" } else { "local variables" };
                    this.r.tcx.sess.create_err(ImportsCannotReferTo { span: ident.span, what }).emit();
```

File: compiler/rustc_resolve/src/diagnostics.rs [hunk 2]
```diff
@@ -1397,12 +1394,12 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
             let source_map = self.tcx.sess.source_map();
             let head_span = source_map.guess_head_span(span);
             if let Ok(head) = source_map.span_to_snippet(head_span) {
-                err.span_suggestion(head_span, "consider adding a derive", format!("#[derive(Default)]\n{head}"), Applicability::MaybeIncorrect);
+                err.subdiagnostic(ConsiderAddingADerive {
+                    span: head_span,
+                    suggestion: format!("#[derive(Default)]\n{head}")
+                });
```
Comment (WaffleLapkin) [inline compiler/rustc_resolve/src/diagnostics.rs:1400]: Why is this using `span_to_snippet`? can't this just be
```rust
ConsiderAddingADerive {
    span: head_span.shrink_to_lo(),
    suggestion: format!("#[derive(Default)]\n")
}
```

**Review by WaffleLapkin — APPROVED** [review]
A few nits, but overall looks good.

----

> For trivial strings is it ever accepted to use `fluent_generated::foo` in a `label` for example? Or is an empty `struct Diagnostic` preferred?

_I think_, an empty struct is preferred, but I'd suggest you ask in the [i18n](https://rust-lang.zulipchat.com/#narrow/stream/336883-i18n) or [t-compiler/wg-diagnostics](
https://rust-lang.zulipchat.com/#narrow/stream/147480-t-compiler.2Fwg-diagnostics) zulip streams.

File: compiler/rustc_resolve/src/diagnostics.rs (hunk 2 above)
Comment (NotStirred) [reply compiler/rustc_resolve/src/diagnostics.rs:1400]: Are you suggesting removing the surrounding `if` entirely?
There is no test for this currently and the else-case does seem unreachable given a sane span

Assuming that, I'm going to remove the second diag struct

Comment (WaffleLapkin) [reply compiler/rustc_resolve/src/diagnostics.rs]: Yes, exactly.

**Review by WaffleLapkin — APPROVED** [review]

**Pull request merged**